File: standards/v1.0/reference/typescript/ushs-client.ts
/**
 * USHS TypeScript Reference Implementation
 * Universal Self-Healing Standard v1.0 Client Library
 */

import { EventEmitter } from 'events';
import WebSocket from 'ws';
import fetch, { RequestInit } from 'node-fetch';

// Enums
export enum Severity {
  CRITICAL = 'critical',
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low'
}

export enum SessionStatus {
  ACTIVE = 'active',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  TIMEOUT = 'timeout'
}

export enum DeploymentStrategy {
  IMMEDIATE = 'immediate',
  CANARY = 'canary',
  BLUE_GREEN = 'blue-green',
  ROLLING = 'rolling'
}

// Interfaces
export interface ErrorSource {
  service: string;
  version?: string;
  environment?: string;
  location?: string;
  hostname?: string;
  containerID?: string;
}

export interface ErrorDetails {
  type: string;
  message: string;
  code?: string;
  stackTrace?: Array<{
    file?: string;
    function?: string;
    line?: number;
    column?: number;
  }>;
  context?: Record<string, any>;
}

export interface ErrorEvent {
  id?: string;
  timestamp?: string;
  severity: Severity;
  source: ErrorSource;
  error: ErrorDetails;
  correlationId?: string;
  userImpact?: {
    affected?: number;
    severity?: 'none' | 'degraded' | 'unavailable';
  };
  metadata?: Record<string, any>;
}

export interface PatchChange {
  type: 'file' | 'configuration' | 'dependency' | 'infrastructure';
  target: string;
  operation?: 'create' | 'update' | 'delete' | 'replace';
  diff?: string;
  content?: string;
  language?: string;
  framework?: string;
  validation?: {
    syntax?: boolean;
    semantics?: boolean;
    tests?: string[];
  };
}

export interface HealingPatch {
  id?: string;
  sessionId: string;
  errorId: string;
  changes: PatchChange[];
  metadata: {
    confidence: number;
    generator: string;
    generatorVersion?: string;
    strategy?: string;
    reasoning?: string;
    alternatives?: Array<{
      id?: string;
      confidence?: number;
      reason?: string;
    }>;
    estimatedImpact?: {
      performance?: 'improved' | 'neutral' | 'degraded';
      reliability?: 'improved' | 'neutral' | 'degraded';
      security?: 'improved' | 'neutral' | 'degraded';
    };
  };
  approvals?: Array<{
    type?: 'automatic' | 'manual' | 'policy';
    approver?: string;
    status?: 'pending' | 'approved' | 'rejected';
    timestamp?: string;
    comment?: string;
  }>;
}

export interface HealingSession {
  id: string;
  startTime: string;
  endTime?: string;
  status: SessionStatus;
  errorId?: string;
  phases: {
    [key: string]: {
      status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped';
      startTime?: string;
      endTime?: string;
      duration?: number;
      result?: any;
      errors?: Array<{
        code?: string;
        message?: string;
        details?: any;
      }>;
    };
  };
}

export interface USHSClientOptions {
  baseUrl: string;
  authToken?: string;
  apiKey?: string;
  verifySsl?: boolean;
  timeout?: number;
}

export interface CloudEvent {
  specversion: string;
  id: string;
  source: string;
  type: string;
  datacontenttype: string;
  time: string;
  subject?: string;
  data: any;
}

/**
 * USHS Client Implementation
 */
export class USHSClient extends EventEmitter {
  private baseUrl: string;
  private authToken?: string;
  private apiKey?: string;
  private verifySsl: boolean;
  private timeout: number;
  private ws?: WebSocket;
  private wsReconnectTimer?: NodeJS.Timeout;
  private wsReconnectAttempts = 0;

  constructor(options: USHSClientOptions) {
    super();
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.authToken = options.authToken;
    this.apiKey = options.apiKey;
    this.verifySsl = options.verifySsl ?? true;
    this.timeout = options.timeout ?? 30000;
  }

  /**
   * Build request headers
   */
  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    };

    if (this.authToken) {
      headers['Authorization'] = `Bearer ${this.authToken}`;
    } else if (this.apiKey) {
      headers['X-API-Key'] = this.apiKey;
    }

    return headers;
  }

  /**
   * Make HTTP request
   */
  private async request<T>(
    method: string,
    path: string,
    data?: any,
    queryParams?: Record<string, any>
  ): Promise<T> {
    let url = `${this.baseUrl}${path}`;
    
    // Add query parameters
    if (queryParams) {
      const params = new URLSearchParams();
      Object.entries(queryParams).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          params.append(key, String(value));
        }
      });
      const queryString = params.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }

    const options: RequestInit = {
      method,
      headers: this.getHeaders(),
      timeout: this.timeout,
    };

    if (data && ['POST', 'PUT', 'PATCH'].includes(method)) {
      options.body = JSON.stringify(data);
    }

    const response = await fetch(url, options);

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`HTTP ${response.status}: ${error}`);
    }

    if (response.status === 204) {
      return {} as T;
    }

    return response.json() as Promise<T>;
  }

  /**
   * Generate UUID v4
   */
  private generateId(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }

  /**
   * Get current ISO timestamp
   */
  private getTimestamp(): string {
    return new Date().toISOString();
  }

  // Error Management

  /**
   * Report a new error
   */
  async reportError(error: ErrorEvent): Promise<{ errorId: string; sessionId: string }> {
    // Add defaults
    if (!error.id) {
      error.id = this.generateId();
    }
    if (!error.timestamp) {
      error.timestamp = this.getTimestamp();
    }

    return this.request<{ errorId: string; sessionId: string }>(
      'POST',
      '/errors',
      error
    );
  }

  /**
   * Get error details
   */
  async getError(errorId: string): Promise<ErrorEvent> {
    return this.request<ErrorEvent>('GET', `/errors/${errorId}`);
  }

  // Session Management

  /**
   * Start a new healing session
   */
  async startSession(
    errorId: string,
    policy?: string,
    priority?: string
  ): Promise<HealingSession> {
    const data: any = { errorId };
    if (policy) data.policy = policy;
    if (priority) data.priority = priority;

    return this.request<HealingSession>('POST', '/sessions', data);
  }

  /**
   * Get session details
   */
  async getSession(sessionId: string): Promise<HealingSession> {
    return this.request<HealingSession>('GET', `/sessions/${sessionId}`);
  }

  /**
   * List healing sessions
   */
  async listSessions(options?: {
    status?: SessionStatus;
    limit?: number;
    offset?: number;
  }): Promise<{
    sessions: HealingSession[];
    total: number;
    limit: number;
    offset: number;
  }> {
    return this.request('GET', '/sessions', undefined, options);
  }

  /**
   * Cancel a healing session
   */
  async cancelSession(sessionId: string): Promise<void> {
    await this.request<void>('DELETE', `/sessions/${sessionId}`);
  }

  // Patch Management

  /**
   * Get patches for a session
   */
  async getSessionPatches(sessionId: string): Promise<HealingPatch[]> {
    return this.request<HealingPatch[]>('GET', `/sessions/${sessionId}/patches`);
  }

  /**
   * Submit a patch for a session
   */
  async submitPatch(sessionId: string, patch: HealingPatch): Promise<HealingPatch> {
    // Add defaults
    if (!patch.id) {
      patch.id = this.generateId();
    }

    return this.request<HealingPatch>('POST', `/sessions/${sessionId}/patches`, patch);
  }

  // Validation

  /**
   * Validate a patch
   */
  async validatePatch(
    patchId: string,
    options?: {
      tests?: string[];
      environment?: string;
    }
  ): Promise<{
    valid: boolean;
    results: Array<{
      test: string;
      passed: boolean;
      duration: number;
      output?: string;
    }>;
  }> {
    return this.request('POST', `/patches/${patchId}/validate`, options);
  }

  // Deployment

  /**
   * Deploy a patch
   */
  async deployPatch(
    patchId: string,
    strategy: DeploymentStrategy,
    environment: string,
    approvals?: Array<{
      approver: string;
      signature: string;
    }>
  ): Promise<{
    deploymentId: string;
    status: string;
    startTime: string;
  }> {
    const data: any = { strategy, environment };
    if (approvals) data.approvals = approvals;

    return this.request('POST', `/patches/${patchId}/deploy`, data);
  }

  // Health Check

  /**
   * Check system health
   */
  async healthCheck(): Promise<{
    status: 'healthy' | 'degraded' | 'unhealthy';
    components: Record<string, {
      status: string;
      message?: string;
    }>;
  }> {
    return this.request('GET', '/health');
  }

  // WebSocket Support

  /**
   * Connect to WebSocket for real-time events
   */
  connectWebSocket(options?: {
    subscribe?: string[];
    session?: string;
    service?: string;
  }): void {
    let wsUrl = this.baseUrl.replace(/^https?:/, 'ws:').replace(/^https:/, 'wss:');
    wsUrl += '/ws';

    // Build query params
    const params: string[] = [];
    if (options?.subscribe) {
      params.push(`subscribe=${options.subscribe.join(',')}`);
    }
    if (options?.session) {
      params.push(`session=${options.session}`);
    }
    if (options?.service) {
      params.push(`service=${options.service}`);
    }
    if (this.apiKey) {
      params.push(`apikey=${this.apiKey}`);
    }

    if (params.length > 0) {
      wsUrl += '?' + params.join('&');
    }

    // Build headers
    const headers: Record<string, string> = {};
    if (this.authToken && !this.apiKey) {
      headers['Authorization'] = `Bearer ${this.authToken}`;
    }

    // Connect
    this.ws = new WebSocket(wsUrl, { headers });

    this.ws.on('open', () => {
      this.wsReconnectAttempts = 0;
      this.emit('ws:connected');
    });

    this.ws.on('message', (data: WebSocket.Data) => {
      try {
        const event = JSON.parse(data.toString()) as CloudEvent;
        this.emit('event', event);
        this.emit(event.type, event);
      } catch (error) {
        this.emit('ws:error', new Error(`Invalid message: ${data}`));
      }
    });

    this.ws.on('error', (error: Error) => {
      this.emit('ws:error', error);
    });

    this.ws.on('close', (code: number, reason: string) => {
      this.emit('ws:disconnected', { code, reason });
      this.attemptReconnect();
    });

    // Ping interval
    const pingInterval = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.ping();
      }
    }, 30000);

    this.ws.on('close', () => {
      clearInterval(pingInterval);
    });
  }

  /**
   * Attempt to reconnect WebSocket
   */
  private attemptReconnect(): void {
    if (this.wsReconnectAttempts >= 10) {
      this.emit('ws:reconnectFailed');
      return;
    }

    const delay = Math.min(1000 * Math.pow(2, this.wsReconnectAttempts), 60000);
    this.wsReconnectAttempts++;

    this.wsReconnectTimer = setTimeout(() => {
      this.emit('ws:reconnecting', { attempt: this.wsReconnectAttempts });
      this.connectWebSocket();
    }, delay);
  }

  /**
   * Subscribe to event types
   */
  async subscribe(eventTypes: string[], filters?: Record<string, any>): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    const command: any = {
      command: 'subscribe',
      eventTypes
    };
    if (filters) {
      command.filters = filters;
    }

    this.ws.send(JSON.stringify(command));
  }

  /**
   * Unsubscribe from event types
   */
  async unsubscribe(eventTypes: string[]): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    this.ws.send(JSON.stringify({
      command: 'unsubscribe',
      eventTypes
    }));
  }

  /**
   * Disconnect WebSocket
   */
  disconnectWebSocket(): void {
    if (this.wsReconnectTimer) {
      clearTimeout(this.wsReconnectTimer);
      this.wsReconnectTimer = undefined;
    }

    if (this.ws) {
      this.ws.close();
      this.ws = undefined;
    }
  }

  /**
   * Clean up resources
   */
  destroy(): void {
    this.disconnectWebSocket();
    this.removeAllListeners();
  }
}

// Example usage
async function exampleUsage() {
  // Create client
  const client = new USHSClient({
    baseUrl: 'https://api.example.com/ushs/v1',
    authToken: 'your-auth-token'
  });

  try {
    // Report an error
    const { errorId, sessionId } = await client.reportError({
      severity: Severity.HIGH,
      source: {
        service: 'api-gateway',
        environment: 'production'
      },
      error: {
        type: 'NullPointerException',
        message: "Cannot read property 'id' of null",
        stackTrace: [
          { file: 'app.js', function: 'getUser', line: 42 }
        ]
      }
    });

    console.log(`Error reported: ${errorId}`);
    console.log(`Session started: ${sessionId}`);

    // Connect WebSocket for real-time updates
    client.connectWebSocket({
      subscribe: ['session', 'patch'],
      session: sessionId
    });

    // Listen for events
    client.on('org.ushs.patch.generated', (event: CloudEvent) => {
      console.log(`Patch generated: ${event.data.patchId}`);
    });

    client.on('org.ushs.session.completed', (event: CloudEvent) => {
      console.log(`Session completed: ${event.data.sessionId}`);
    });

    // Check session status
    const session = await client.getSession(sessionId);
    console.log(`Session status: ${session.status}`);

    // Wait for healing to complete
    await new Promise(resolve => setTimeout(resolve, 30000));

  } finally {
    // Clean up
    client.destroy();
  }
}

// Run example if this is the main module
if (require.main === module) {
  exampleUsage().catch(console.error);
}

export default USHSClient;

File: standards/v1.0/reference/python/ushs_client.py
"""
USHS Python Reference Implementation
Universal Self-Healing Standard v1.0 Client Library
"""

import asyncio
import json
import logging
import uuid
from dataclasses import asdict, dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Callable, Dict, List, Optional
from urllib.parse import urljoin

import aiohttp
import websockets


class Severity(Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class SessionStatus(Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    TIMEOUT = "timeout"


class DeploymentStrategy(Enum):
    IMMEDIATE = "immediate"
    CANARY = "canary"
    BLUE_GREEN = "blue-green"
    ROLLING = "rolling"


@dataclass
class ErrorEvent:
    """USHS Error Event"""

    severity: Severity
    service: str
    type: str
    message: str
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: str = field(default_factory=lambda: datetime.utcnow().isoformat() + "Z")
    version: Optional[str] = None
    environment: Optional[str] = None
    location: Optional[str] = None
    hostname: Optional[str] = None
    container_id: Optional[str] = None
    code: Optional[str] = None
    stack_trace: Optional[List[Dict[str, Any]]] = None
    context: Optional[Dict[str, Any]] = None
    correlation_id: Optional[str] = None
    user_impact: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to USHS-compliant dictionary"""
        data = {
            "id": self.id,
            "timestamp": self.timestamp,
            "severity": self.severity.value,
            "source": {"service": self.service},
            "error": {"type": self.type, "message": self.message},
        }

        # Add optional source fields
        if self.version:
            data["source"]["version"] = self.version
        if self.environment:
            data["source"]["environment"] = self.environment
        if self.location:
            data["source"]["location"] = self.location
        if self.hostname:
            data["source"]["hostname"] = self.hostname
        if self.container_id:
            data["source"]["containerID"] = self.container_id

        # Add optional error fields
        if self.code:
            data["error"]["code"] = self.code
        if self.stack_trace:
            data["error"]["stackTrace"] = self.stack_trace
        if self.context:
            data["error"]["context"] = self.context

        # Add optional top-level fields
        if self.correlation_id:
            data["correlationId"] = self.correlation_id
        if self.user_impact:
            data["userImpact"] = self.user_impact
        if self.metadata:
            data["metadata"] = self.metadata

        return data


@dataclass
class HealingPatch:
    """USHS Healing Patch"""

    session_id: str
    error_id: str
    changes: List[Dict[str, Any]]
    confidence: float
    generator: str
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    generator_version: Optional[str] = None
    strategy: Optional[str] = None
    reasoning: Optional[str] = None
    alternatives: Optional[List[Dict[str, Any]]] = None
    estimated_impact: Optional[Dict[str, Any]] = None
    approvals: Optional[List[Dict[str, Any]]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to USHS-compliant dictionary"""
        data = {
            "id": self.id,
            "sessionId": self.session_id,
            "errorId": self.error_id,
            "changes": self.changes,
            "metadata": {"confidence": self.confidence, "generator": self.generator},
        }

        # Add optional metadata fields
        if self.generator_version:
            data["metadata"]["generatorVersion"] = self.generator_version
        if self.strategy:
            data["metadata"]["strategy"] = self.strategy
        if self.reasoning:
            data["metadata"]["reasoning"] = self.reasoning
        if self.alternatives:
            data["metadata"]["alternatives"] = self.alternatives
        if self.estimated_impact:
            data["metadata"]["estimatedImpact"] = self.estimated_impact

        # Add optional top-level fields
        if self.approvals:
            data["approvals"] = self.approvals

        return data


class USHSClient:
    """USHS Reference Client Implementation"""

    def __init__(
        self,
        base_url: str,
        auth_token: Optional[str] = None,
        api_key: Optional[str] = None,
        verify_ssl: bool = True,
    ):
        self.base_url = base_url.rstrip("/")
        self.auth_token = auth_token
        self.api_key = api_key
        self.verify_ssl = verify_ssl
        self.logger = logging.getLogger(__name__)
        self._session: Optional[aiohttp.ClientSession] = None
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._ws_handlers: Dict[str, List[Callable]] = {}

    async def __aenter__(self):
        await self._ensure_session()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def _ensure_session(self):
        """Ensure HTTP session exists"""
        if not self._session:
            headers = {}
            if self.auth_token:
                headers["Authorization"] = f"Bearer {self.auth_token}"
            elif self.api_key:
                headers["X-API-Key"] = self.api_key

            connector = aiohttp.TCPConnector(ssl=self.verify_ssl)
            self._session = aiohttp.ClientSession(headers=headers, connector=connector)

    async def close(self):
        """Close all connections"""
        if self._ws:
            await self._ws.close()
            self._ws = None
        if self._session:
            await self._session.close()
            self._session = None

    # Error Management

    async def report_error(self, error: ErrorEvent) -> Dict[str, str]:
        """Report a new error to the healing system"""
        await self._ensure_session()

        url = urljoin(self.base_url, "/errors")
        async with self._session.post(url, json=error.to_dict()) as resp:
            resp.raise_for_status()
            return await resp.json()

    async def get_error(self, error_id: str) -> Dict[str, Any]:
        """Get error details"""
        await self._ensure_session()

        url = urljoin(self.base_url, f"/errors/{error_id}")
        async with self._session.get(url) as resp:
            resp.raise_for_status()
            return await resp.json()

    # Session Management

    async def start_session(
        self,
        error_id: str,
        policy: Optional[str] = None,
        priority: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Start a new healing session"""
        await self._ensure_session()

        data = {"errorId": error_id}
        if policy:
            data["policy"] = policy
        if priority:
            data["priority"] = priority

        url = urljoin(self.base_url, "/sessions")
        async with self._session.post(url, json=data) as resp:
            resp.raise_for_status()
            return await resp.json()

    async def get_session(self, session_id: str) -> Dict[str, Any]:
        """Get session details"""
        await self._ensure_session()

        url = urljoin(self.base_url, f"/sessions/{session_id}")
        async with self._session.get(url) as resp:
            resp.raise_for_status()
            return await resp.json()

    async def list_sessions(
        self, status: Optional[SessionStatus] = None, limit: int = 20, offset: int = 0
    ) -> Dict[str, Any]:
        """List healing sessions"""
        await self._ensure_session()

        params = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status.value

        url = urljoin(self.base_url, "/sessions")
        async with self._session.get(url, params=params) as resp:
            resp.raise_for_status()
            return await resp.json()

    async def cancel_session(self, session_id: str) -> None:
        """Cancel a healing session"""
        await self._ensure_session()

        url = urljoin(self.base_url, f"/sessions/{session_id}")
        async with self._session.delete(url) as resp:
            resp.raise_for_status()

    # Patch Management

    async def get_session_patches(self, session_id: str) -> List[Dict[str, Any]]:
        """Get patches for a session"""
        await self._ensure_session()

        url = urljoin(self.base_url, f"/sessions/{session_id}/patches")
        async with self._session.get(url) as resp:
            resp.raise_for_status()
            return await resp.json()

    async def submit_patch(
        self, session_id: str, patch: HealingPatch
    ) -> Dict[str, Any]:
        """Submit a patch for a session"""
        await self._ensure_session()

        url = urljoin(self.base_url, f"/sessions/{session_id}/patches")
        async with self._session.post(url, json=patch.to_dict()) as resp:
            resp.raise_for_status()
            return await resp.json()

    # Validation

    async def validate_patch(
        self,
        patch_id: str,
        tests: Optional[List[str]] = None,
        environment: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Validate a patch"""
        await self._ensure_session()

        data = {}
        if tests:
            data["tests"] = tests
        if environment:
            data["environment"] = environment

        url = urljoin(self.base_url, f"/patches/{patch_id}/validate")
        async with self._session.post(url, json=data) as resp:
            resp.raise_for_status()
            return await resp.json()

    # Deployment

    async def deploy_patch(
        self,
        patch_id: str,
        strategy: DeploymentStrategy,
        environment: str,
        approvals: Optional[List[Dict[str, str]]] = None,
    ) -> Dict[str, Any]:
        """Deploy a patch"""
        await self._ensure_session()

        data = {"strategy": strategy.value, "environment": environment}
        if approvals:
            data["approvals"] = approvals

        url = urljoin(self.base_url, f"/patches/{patch_id}/deploy")
        async with self._session.post(url, json=data) as resp:
            resp.raise_for_status()
            return await resp.json()

    # WebSocket Support

    async def connect_websocket(
        self,
        subscribe: Optional[List[str]] = None,
        session: Optional[str] = None,
        service: Optional[str] = None,
    ):
        """Connect to WebSocket for real-time events"""
        ws_url = self.base_url.replace("https://", "wss://").replace("http://", "ws://")
        ws_url = urljoin(ws_url, "/ws")

        # Build query params
        params = []
        if subscribe:
            params.append(f"subscribe={','.join(subscribe)}")
        if session:
            params.append(f"session={session}")
        if service:
            params.append(f"service={service}")
        if self.api_key:
            params.append(f"apikey={self.api_key}")

        if params:
            ws_url += "?" + "&".join(params)

        # Build headers
        headers = {}
        if self.auth_token and not self.api_key:
            headers["Authorization"] = f"Bearer {self.auth_token}"

        self._ws = await websockets.connect(ws_url, extra_headers=headers)

        # Start message handler
        asyncio.create_task(self._handle_ws_messages())

    async def _handle_ws_messages(self):
        """Handle incoming WebSocket messages"""
        try:
            async for message in self._ws:
                try:
                    event = json.loads(message)
                    event_type = event.get("type", "")

                    # Call registered handlers
                    handlers = self._ws_handlers.get(event_type, [])
                    handlers.extend(self._ws_handlers.get("*", []))

                    for handler in handlers:
                        try:
                            await handler(event)
                        except Exception as e:
                            self.logger.error(f"Error in WebSocket handler: {e}")

                except json.JSONDecodeError:
                    self.logger.error(f"Invalid WebSocket message: {message}")

        except websockets.exceptions.ConnectionClosed:
            self.logger.info("WebSocket connection closed")
        except Exception as e:
            self.logger.error(f"WebSocket error: {e}")

    def on_event(self, event_type: str):
        """Decorator to register WebSocket event handlers"""

        def decorator(func):
            if event_type not in self._ws_handlers:
                self._ws_handlers[event_type] = []
            self._ws_handlers[event_type].append(func)
            return func

        return decorator

    async def subscribe(
        self, event_types: List[str], filters: Optional[Dict[str, Any]] = None
    ):
        """Subscribe to additional event types"""
        if not self._ws:
            raise RuntimeError("WebSocket not connected")

        command = {"command": "subscribe", "eventTypes": event_types}
        if filters:
            command["filters"] = filters

        await self._ws.send(json.dumps(command))

    async def unsubscribe(self, event_types: List[str]):
        """Unsubscribe from event types"""
        if not self._ws:
            raise RuntimeError("WebSocket not connected")

        command = {"command": "unsubscribe", "eventTypes": event_types}

        await self._ws.send(json.dumps(command))

    # Health Check

    async def health_check(self) -> Dict[str, Any]:
        """Check system health"""
        await self._ensure_session()

        url = urljoin(self.base_url, "/health")
        async with self._session.get(url) as resp:
            resp.raise_for_status()
            return await resp.json()


# Example usage
async def example_usage():
    """Example of using the USHS client"""

    # Create client
    async with USHSClient(
        base_url="https://api.example.com/ushs/v1", auth_token="your-auth-token"
    ) as client:

        # Report an error
        error = ErrorEvent(
            severity=Severity.HIGH,
            service="api-gateway",
            type="NullPointerException",
            message="Cannot read property 'id' of null",
            environment="production",
            stack_trace=[{"file": "app.js", "function": "getUser", "line": 42}],
        )

        result = await client.report_error(error)
        error_id = result["errorId"]
        session_id = result["sessionId"]

        # Check session status
        session = await client.get_session(session_id)
        print(f"Session status: {session['status']}")

        # Connect WebSocket for real-time updates
        await client.connect_websocket(subscribe=["session", "patch"])

        # Register event handlers
        @client.on_event("org.ushs.patch.generated")
        async def on_patch_generated(event):
            print(f"Patch generated: {event['data']['patchId']}")

        @client.on_event("org.ushs.session.completed")
        async def on_session_completed(event):
            print(f"Session completed: {event['data']['sessionId']}")

        # Wait for healing to complete
        await asyncio.sleep(30)


if __name__ == "__main__":
    asyncio.run(example_usage())
